File: android/src/com/rsb/splyt/SplytWrapper.java
package com.rsb.splyt;

import java.lang.reflect.Type;
import java.util.Map;

import com.unity3d.player.UnityPlayer;

import com.rsb.gson.Gson;
import com.rsb.gson.reflect.TypeToken;
import com.rsb.splyt.Splyt.Core;
import com.rsb.splyt.Splyt.Instrumentation;
import com.rsb.splyt.Splyt.Tuning;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;

public class SplytWrapper {
    static private Map<String, Object> toDictionary(String string) {
        Gson gson = new Gson();
        Type bagType = new TypeToken<Map<String, Object>>() {}.getType();
        Map<String, Object> dic = gson.fromJson(string, bagType);
        return dic;
    }
    
    static private Splyt.Instrumentation.Transaction toTransaction(String category, String transactionId, String txnProperties) {
        return Instrumentation.Transaction(category, transactionId).setProperties(toDictionary(txnProperties));
    }
    
    static public String splyt_core_getuserid() {
        return Core.getUserId();
    }
    
    static public String splyt_core_getdeviceid() {
        return Core.getDeviceId();
    }
    
    static public void splyt_core_init(Activity activity, String customerId, String userId, String userProperties, String deviceId, String deviceProperties, int reqTimeout, String host, boolean logEnabled, String sdkNamePre, String sdkVersion, String notificationHost, int notificationSmallIcon, boolean notificationAlwaysPost, boolean notificationDisableAutoClear, String callbackObj) {
        Log.i("Unity", "in SplytWrapper.splyt_core_init");
        Core.EntityInfo userInfo = Core.createUserInfo(userId).setProperties(toDictionary(userProperties));
        Core.EntityInfo deviceInfo = Core.createDeviceInfo().overrideId(deviceId).setProperties(toDictionary(deviceProperties));

        Core.InitParams initParams = Core.createInitParams(activity, customerId)
            .setUserInfo(userInfo)
            .setDeviceInfo(deviceInfo)
            .setRequestTimeout(reqTimeout)
            .setHost(host)
            .setLogEnabled(logEnabled)
            .setSDKName(sdkNamePre + "-android")
            .setSDKVersion(sdkVersion);

        final String cbObj = callbackObj;

        initParams.Notification
            .setHost(notificationHost)
            .setSmallIcon(notificationSmallIcon)
            .setAlwaysPost(notificationAlwaysPost)
            .setDisableAutoClear(notificationDisableAutoClear)
            .setReceivedListener(new SplytNotificationReceivedListener() {
                @Override
                public void onReceived(Bundle info, boolean wasLaunchedBy) {
                    android.util.Log.i("Unity", "NotificationListener.onReceived: " + wasLaunchedBy + info);
                    String unityMessage = (wasLaunchedBy ? "1" : "0");
                    String message = info.getString("message");
                    if (message != null) {
                        unityMessage += message;
                    }
                    UnityPlayer.UnitySendMessage(cbObj, "onSplytNotificationReceived", unityMessage);
                }
            });
        
        Log.i("Unity", "calling Splyt.Core.init");
        Splyt.Core.init(initParams, new SplytListener() {
            @Override
            public void onComplete(SplytError error) {
                UnityPlayer.UnitySendMessage(cbObj, "onSplytInitComplete", String.valueOf(error.getValue()));
            }
        });
    }
    
    static public void splyt_core_registeruser(String userId, String userProperties, String callbackObj) {
        Core.EntityInfo userInfo = Core.createUserInfo(userId).setProperties(toDictionary(userProperties));
        
        final String cbObj = callbackObj;
        Core.registerUser(userInfo, new SplytListener() {
            @Override
            public void onComplete(SplytError error) {
                UnityPlayer.UnitySendMessage(cbObj, "onSplytRegisterUserComplete", String.valueOf(error.getValue()));
            }
        });
    }
    
    static public String splyt_core_setactiveuser(String userId) {
        SplytError error = Core.setActiveUser(userId);
        return String.valueOf(error.getValue());
    }

    static public String splyt_core_clearactiveuser() {
        SplytError error = Core.clearActiveUser();
        return String.valueOf(error.getValue());
    }

    static public void splyt_core_pause(Object nullObj) {
        Core.pause();
    }

    static public void splyt_core_resume(Object nullObj) {
        Core.resume();
    }

    static public void splyt_transaction_begin(String category, String transactionId, String txnProperties, double timeout, int timeoutMode) {
        String mode = (timeoutMode == 0) ? "TXN" : "ANY";
        
        toTransaction(category, transactionId, txnProperties).begin(mode, timeout);
    }

    static public void splyt_transaction_update(String category, String transactionId, String txnProperties, int progress) {
        toTransaction(category, transactionId, txnProperties).update(progress);
    }

    static public void splyt_transaction_end(String category, String transactionId, String txnProperties, String result) {
        toTransaction(category, transactionId, txnProperties).end(result);
    }

    static public void splyt_transaction_beginandend(String category, String transactionId, String txnProperties, String result) {
        toTransaction(category, transactionId, txnProperties).beginAndEnd(result);
    }

    static public void splyt_instrumentation_updatedevicestate(String deviceProperties) {
        Instrumentation.updateDeviceState(toDictionary(deviceProperties));
    }

    static public void splyt_instrumentation_updateuserstate(String userProperties) {
        Instrumentation.updateUserState(toDictionary(userProperties));
    }

    static public void splyt_instrumentation_updatecollection(String name, double balance, double balanceModification, boolean isCurrency) {
        Instrumentation.updateCollection(name, balance, balanceModification, isCurrency);
    }

    static public void splyt_tuning_refresh(String callbackObj) {
        final String cbObj = callbackObj;
        Tuning.refresh(new SplytListener() {
            @Override
            public void onComplete(SplytError error) {
                UnityPlayer.UnitySendMessage(cbObj, "onSplytRefreshComplete", String.valueOf(error.getValue()));
            }
        });
    }

    static public String splyt_tuning_getvar(String varName, String defaultValue) {
        return Tuning.getVar(varName, defaultValue);
    }

    static public String splyt_util_getvalidcurrencystring(String currency) {
        return Util.getValidCurrencyString(currency);
    }

}

File: samples/BubblePop/Assets/Plugins/iOS/SplytWrapper.m
#import <Splyt/Splyt.h>

// By default mono marshals all .Net strings as UTF-8 C style strings, but ObjC wants NSString
NSString* toNSString(const char* string) {
    if(string)
        return [NSString stringWithUTF8String:string];
    else
        return nil;
}

NSDictionary* toDictionary(const char* string) {
    NSError* error;
    id properties = [NSJSONSerialization JSONObjectWithData:[NSData dataWithBytesNoCopy:(void*)string length:strlen(string) freeWhenDone:NO] options:0 error:&error];
    if(properties && [properties isKindOfClass:[NSDictionary class]])
        return properties;
    return nil;
}

SplytTransaction* toTransaction(const char* category, const char* transactionId, const char* txnProperties) {
    return [[Splyt Instrumentation] Transaction:toNSString(category) withId:toNSString(transactionId) andInitBlock:^(SplytTransaction *t) {
        [t setProperties:toDictionary(txnProperties)];
    }];
}

// By default mono string marshaler creates .Net string for returned UTF-8 C string
// and calls free for returned value, thus returned strings should be allocated on heap
char* toHeapString(NSString* str) {
    if(nil == str)
        return NULL;
    
    const char* string = [str UTF8String];
    
    char* res = (char*)malloc(strlen(string) + 1);
	strcpy(res, string);
	return res;
}

@interface SplytInitParams ()
// convince the compiler that I know about these, even though they aren't REALLY exposed
@property NSString* SDKName;
@property NSString* SDKVersion;
@end

char* splyt_core_getuserid() {
    NSString* userId = [[Splyt Core] userId];
    return toHeapString(userId);
}
char* splyt_core_getdeviceid() {
    NSString* deviceId = [[Splyt Core] deviceId];
    return toHeapString(deviceId);
}

void splyt_core_init(const char* customerId, const char* userId, const char* userProperties, const char* deviceId, const char* deviceProperties, int reqTimeout, const char* host, bool logEnabled, const char* sdkNamePre, const char* sdkVersion, const char* notificationHost, bool notificationDisableAutoClear, const char* callbackObj) {
    SplytEntityInfo* userInfo = [SplytEntityInfo createUserInfo:toNSString(userId) withInitBlock:^(SplytEntityInfo *u) {
        [u setProperties:toDictionary(userProperties)];
    }];
    SplytEntityInfo* deviceInfo = [SplytEntityInfo createDeviceInfoWithInitBlock:^(SplytEntityInfo *d) {
        if(deviceId)
            [d overrideId:toNSString(deviceId)];
        
        [d setProperties:toDictionary(deviceProperties)];
    }];

    // need to wrap this in an NSString for the async journey
    NSString* cbStr = toNSString(callbackObj);

    SplytInitParams* initParams = [SplytInitParams createWithCustomerId:toNSString(customerId) andInitBlock:^(SplytInitParams *init) {
        init.userInfo = userInfo;
        init.deviceInfo = deviceInfo;
        init.requestTimeout = reqTimeout;
        init.host = toNSString(host);
        init.logEnabled = logEnabled;
        init.SDKName = [toNSString(sdkNamePre) stringByAppendingString:@"-ios"];
        init.SDKVersion = toNSString(sdkVersion);
        init.notification.host = toNSString(notificationHost);
        init.notification.disableAutoClear = notificationDisableAutoClear;
        init.notification.receivedCallback = ^(NSDictionary* info, BOOL wasLaunchedBy) {
            NSString *message = @"";
            id rawMessage = [info objectForKey:@"message"];
            if ([rawMessage isKindOfClass:[NSString class]]) {
                message = (NSString *)rawMessage;
            }
            NSString *unityMessage = [NSString stringWithFormat:@"%d%@", wasLaunchedBy ? 1 : 0, message];
            UnitySendMessage([cbStr UTF8String], "onSplytNotificationReceived", [unityMessage cString]);
        };
    }];
    
    [[Splyt Core] init:initParams andThen:^(SplytError error) {
        UnitySendMessage([cbStr UTF8String], "onSplytInitComplete", [[@(error) description] UTF8String]);
    }];
}

void splyt_core_registeruser(const char* userId, const char* userProperties, const char* callbackObj) {
    SplytEntityInfo* userInfo = [SplytEntityInfo createUserInfo:toNSString(userId) withInitBlock:^(SplytEntityInfo *u) {
        [u setProperties:toDictionary(userProperties)];
    }];
    
    // need to wrap this in an NSString for the async journey
    NSString* cbStr = toNSString(callbackObj);
    
    [[Splyt Core] registerUser:userInfo andThen:^(SplytError error) {
        UnitySendMessage([cbStr UTF8String], "onSplytRegisterUserComplete", [[@(error) description] UTF8String]);
    }];
}

char* splyt_core_setactiveuser(const char* userId) {
    SplytError error = [[Splyt Core] setActiveUser:toNSString(userId)];
    return toHeapString([@(error) description]);
}

char* splyt_core_clearactiveuser() {
    SplytError error = [[Splyt Core] clearActiveUser];
    return toHeapString([@(error) description]);
}

void splyt_core_pause() {
    [[Splyt Core] pause];
}

void splyt_core_resume() {
    [[Splyt Core] resume];
}

void splyt_transaction_begin(const char* category, const char* transactionId, const char* txnProperties, double timeout, int timeoutMode) {
    SplytTimeoutMode mode = (SplytTimeoutMode) timeoutMode;
    
    [toTransaction(category, transactionId, txnProperties) beginWithTimeout:timeout andMode:mode];
}

void splyt_transaction_update(const char* category, const char* transactionId, const char* txnProperties, int progress) {
    [toTransaction(category, transactionId, txnProperties) updateAtProgress:progress];
}

void splyt_transaction_end(const char* category, const char* transactionId, const char* txnProperties, const char* result) {
    [toTransaction(category, transactionId, txnProperties) endWithResult:toNSString(result)];
}

void splyt_transaction_beginandend(const char* category, const char* transactionId, const char* txnProperties, const char* result) {
    [toTransaction(category, transactionId, txnProperties) beginAndEndWithResult:toNSString(result)];
}

void splyt_instrumentation_updatedevicestate(const char* deviceProperties) {
    [[Splyt Instrumentation] updateDeviceState:toDictionary(deviceProperties)];
}

void splyt_instrumentation_updateuserstate(const char* userProperties) {
    [[Splyt Instrumentation] updateUserState:toDictionary(userProperties)];
}

void splyt_instrumentation_updatecollection(const char* name, double balance, double balanceModification, bool isCurrency) {
    [[Splyt Instrumentation] updateCollection:toNSString(name) toBalance:@(balance) byAdding:@(balanceModification) andTreatAsCurrency:isCurrency];
}

void splyt_tuning_refresh(const char* callbackObj) {
    NSString* cbStr = toNSString(callbackObj);
    [[Splyt Tuning] refreshAndThen:^(SplytError error) {
        UnitySendMessage([cbStr UTF8String], "onSplytRefreshComplete", [[@(error) description] UTF8String]);
    }];
}

char* splyt_tuning_getvar(const char* varName, const char* defaultValue) {
    id val = [[Splyt Tuning] getVar:toNSString(varName) orDefaultTo:toNSString(defaultValue)];
    
    return toHeapString([val description]);
}

// Pull in the "hidden" utility function
NSString* SplytUtil_getValidCurrencyString(NSString* currency);
char* splyt_util_getvalidcurrencystring(const char* currency) {
    return toHeapString(SplytUtil_getValidCurrencyString(toNSString(currency)));
}

